### PR TITLE
Don't nest the sync within 'viewRemoteBranch'

### PR DESCRIPTION
**Pull request by ChrisPenner**
When the sync is inside the viewRemoteBranch call we end up with
an existing open connection to the database when we try to switch modes
to DELETE mode, and this means the WAL files aren't cleaned up properly.

NOTE: This PR isn't ready to merge, it's in a broken state, we need to come up with a solution to the 'leaking codebase connections' problem, this just demonstrates both the issue I'm trying to solve, and the problem that solving it creates.

## Overview

The current code nests a `sync` action INSIDE the `viewRemoteBranch` action block. both of these actions open a separate connection to the same remote codebase. When the `sync` action tries to push, it first switches journal modes to `DELETE`, and expects sqlite to clean up the WAL files, BUT since there's STILL an open connection in the containing `viewRemoteBranch` block it can't.

I believe this is what's causing me trouble in the #2638 PR.

## Implementation notes

I just moved the sync block OUTSIDE of the scope of the `viewRemoteBranch` action.

## Test coverage

This will be more thoroughly tested in #2638 , but we should probably try to prevent this from happening by accident using some form of type-level thing in the future I think.

## Loose ends

* We should likely investigate better ways to prevent accidental connection nesting like this in the future, or at the very least we should always open a given codebase using the same connection type.
* This version of the code has an additional problem that it the remote branch outside of its codebase scope (of viewRemoteBranch), which I don't believe is an issue in this particular case because I don't think sync ever actually reads the history of the branch, and thus doesn't use the 'm' here, but it would be much safer if we had a type-level guarantee of this.

Some potential avenues we can look into for better safety:
* Introduce a ShallowBranch type which doesn't have any database connections embedded, then generalize the `m` on actions which don't actually look at history or patches etc.

```haskell
data VoidF a
type ShallowBranch = Branch VoidF
```

* Scope the codebase monad similar to how `ST` has an `s` parameter to prevent people from leaking connections outside of their scope. E.g. we would add a phantom `s` param to Branch which would bind each branch to the scope of its parent codebase.

```
viewRemoteBranch ::
  MonadIO m =>
  Codebase m v a ->
  ReadRemoteNamespace ->
  (forall s. Branch s m -> m r) ->
  m (Either GitError r)
```

This trick is effective in preventing leaks, but can be a bit confusing for the uninitiated. [Here's an article](https://thoughtsfactorial.com/posts/2021-08-07-the-st-trick/) if anyone is unfamiliar.

This PR doesn't attempt to solve either of these issues.

### CODE REVIEW

File: unison-cli/src/Unison/Codebase/Editor/HandleInput.hs [hunk 1]
```diff
@@ -1879,17 +1879,22 @@ doPushRemoteBranch repo localPath syncMode remoteTarget = do
...
-                  Right () -> respond Success
+          maybeNewRemoteBranch <-
+            ExceptT . viewRemoteBranch (writeToRead repo, Nothing, Path.empty) $ \remoteRoot -> do
+              let -- We don't merge `sourceBranch` with `remoteBranch`, we just replace it. This push will be rejected if this
+                  -- rewinds time or misses any new updates in the remote branch that aren't in `sourceBranch` already.
+                  f remoteBranch = if shouldPushTo pushBehavior remoteBranch then Just sourceBranch else Nothing
+              pure $ Branch.modifyAtM remotePath f remoteRoot
+          case maybeNewRemoteBranch of
+            Nothing -> respond (RefusedToPush pushBehavior)
+            Just newRemoteRoot -> do
```
Comment (ChrisPenner) [inline unison-cli/src/Unison/Codebase/Editor/HandleInput.hs:1890]: Note that this remote branch contains `m`'s with closed connections. See my on the bottom of the PR description for ideas on potentially fixing this.

**Comment by ChrisPenner** [comment]
I think I can get around the issues here by changing `pushGitBranch` in the SqliteCodebase layer to accept a path and move some of the logic from HandleInput into there, it kind of seems like something the codebase should know about anyways.

**Comment by aryairani** [comment]
> When the sync action tries to push, it first switches journal modes to DELETE, and expects sqlite to clean up the WAL files, BUT since there's STILL an open connection in the containing viewRemoteBranch block it can't.

My understanding is that this shouldn't be a problem, that you could switch the journal mode to DELETE while there are multiple connections open.  Is that not true?

**Comment by ChrisPenner** [comment]
@aryairani 
> My understanding is that this shouldn't be a problem, that you could switch the journal mode to DELETE while there are multiple connections open. Is that not true?

I'd be really interested in what articles or documentation you've read on the topic that give you that understanding, as I've had a tough time finding any resources at all from sqlite about what happens when you switch journal modes with multiple connections open, as far as I can tell the expected behaviour isn't defined anywhere, so my understanding simply comes from what throws errors and what doesn't, and right now this is throwing an error (it doesn't throw an error on trunk because every git operation force deletes any extra database files, but this just masks the problem it doesn't avoid it)

From all the resources I can find it really really seems like we shouldn't be switching journal modes in in the middle of multiple connections doing things.

Regardless, I'm not saying we need to switch to WAL mode here either, there are many possible solutions.

**Comment by aryairani** [comment]
> From all the resources I can find it really really seems like we shouldn't be switching journal modes in in the middle of multiple connections doing things.

Would you link some of these to the thread? I'll try to do the same.

**Comment by ChrisPenner** [comment]
This needs to be rewritten using everything we learned in #2638

**Pull request closed without merging**